packages/hardhat/contracts/CoinEngine.sol

```diff
@@ -51,52 +61,64 @@ contract MyUSDEngine is Ownable {
         i_staking = Staking(stakingAddress);
     }
 
-    function setInterestRate(uint256 newRate) external onlyOwner {
-        if (newRate > 2000) revert Engine__InvalidInterestRate(); // Max 20%
+    function setBorrowRate(uint256 newRate) external onlyOwner {
+        _accrueInterest();
+        borrowRate = newRate;
+        emit BorrowRateUpdated(newRate);
+    }
+
+    function setSavingsRate(uint256 newRate) external onlyOwner {
+        if (newRate > borrowRate) revert Engine__InvalidSavingsRate();
         _accrueInterest();
-        interestRate = newRate;
-        emit InterestRateUpdated(newRate);
+        myUSDSavingsRate = newRate;
+        emit SavingsRateUpdated(newRate);
     }
 
     function _accrueInterest() internal {
-        if (totalDebt == 0) {
+        if (totalDebtShares == 0) {
             lastUpdateTime = block.timestamp;
             return;
         }
 
         uint256 timeElapsed = block.timestamp - lastUpdateTime;
         if (timeElapsed == 0) return;
 
-        uint256 interest = (totalDebt * interestRate * timeElapsed) / (SECONDS_PER_YEAR * 10000);
+        // Calculate total debt value
+        uint256 totalDebtValue = (totalDebtShares * debtExchangeRate) / PRECISION;
+
+        // Calculate interest based on total debt value
+        uint256 interest = (totalDebtValue * borrowRate * timeElapsed) / (SECONDS_PER_YEAR * 10000);
+
         if (interest > 0) {
-            totalDebt += interest;
+            // Update exchange rate to reflect new value
+            debtExchangeRate += (interest * PRECISION) / totalDebtShares;
 
             // Mint interest to the staking contract
             i_myUSD.mintTo(address(i_staking), interest);
 
             // Update the staking contract's exchange rate to reflect the new interest
-            i_staking.setInterestRate(interestRate);
+            i_staking.setInterestRate(myUSDSavingsRate);
 
             emit InterestAccrued(interest);
         }
         lastUpdateTime = block.timestamp;
     }
 
-    function _updateUserDebt(address user) internal {
-        if (s_userMinted[user] == 0) {
-            s_userLastUpdateTime[user] = block.timestamp;
-            return;
-        }
+    // Calculate the current debt value for a user, including accrued interest
+    function _getCurrentDebtValue(address user) internal view returns (uint256) {
```

Comment:
This is perfect. The way you are calculating the "baseDebt" and then adding the amount accrued with time is great.